src/ws/session.c

```diff
@@ -568,35 +575,35 @@ store_krb_credentials (gss_cred_id_t creds, uid_t uid, gid_t gid)
 
   assert (geteuid () == uid && getegid() == gid);
 
-  krb5_context k5;
-  int res = krb5_init_context (&k5);
-  if (res == 0)
-    {
-      gss_key_value_set_desc store;
-      struct gss_key_value_element_struct element;
-      OM_uint32 major, minor;
-
-      store.count = 1;
-      store.elements = &element;
-      element.key = "ccache";
-      element.value = krb5_cc_default_name (k5);
-
-      debug ("storing kerberos credentials in session: %s", element.value);
+  /* The ccache path needs to be unique per cockpit session; as cockpit-session runs throughout the
+   * lifetime of sessions, our pid is unique. We expect the cache to be cleaned up at the end, but
+   * if not, and the PID gets recycled, this just overwrites the old obsolete one, which is good. */
+  asprintfx (&ccache, "FILE:/run/user/%u/cockpit-session-%u.ccache", uid, getpid ());
+  debug ("storing kerberos credentials in session: %s", ccache);
 
-      major = gss_store_cred_into (&minor, creds, GSS_C_INITIATE, GSS_C_NULL_OID, 1, 1, &store, NULL, NULL);
-      if (GSS_ERROR (major))
-        warnx ("couldn't store gssapi credentials: %s", gssapi_strerror (GSS_C_NO_OID, major, minor));
-
-      krb5_free_context (k5);
-    }
-  else
-    {
-      warnx ("couldn't initialize kerberos context: %s", krb5_get_error_message (NULL, res));
-    }
+  store.count = 1;
+  store.elements = &element;
+  element.key = "ccache";
+  element.value = ccache;
 
+  major = gss_store_cred_into (&minor, creds, GSS_C_INITIATE, GSS_C_NULL_OID, 1, 1, &store, NULL, NULL);
+  if (GSS_ERROR (major))
+    warnx ("couldn't store gssapi credentials: %s", gssapi_strerror (GSS_C_NO_OID, major, minor));
 
   if (was_root && (setresuid (0, 0, 0) != 0 || setresgid (0, 0, 0) != 0))
     err (127, "Unable to restore permissions after storing gss credentials");
+
+  return ccache;
+}
+
+static void
+release_krb_credentials (char *ccache)
+{
+  /* strip off FILE: prefix for deleting */
+  assert (strncmp (ccache, "FILE:", 5) == 0);
+  if (unlink (ccache + 5) != 0)
+    warn ("couldn't clean up kerberos ticket cache %s", ccache);
+  free (ccache);
```

Comment:
Since you store the string in a static, the create function should probably return `const` and you should definitely not free it here while there's still a pointer around for it.

... or you could just remove the `static` ;)

Reply:
Having a `const` return and not freeing it was my original approach which you didn't like, so I suppose that's a no-go. You really want me to lose that assertion, don't you :grin: (FAOD, I actually *want* this to be a singleton)

Reply:
OK, pushed -- it now does the same thing as before, except with less performance, more ways to fail, and fewer assertions :man_shrugging:

Reply:
... these are a few of my favorite things! 🎶